### PR TITLE
[FIXED JENKINS-29876] CheckForNull job in ReverseBuildTrigger

### PR DESCRIPTION
**Pull request by KostyaSha**
https://issues.jenkins-ci.org/browse/JENKINS-29876

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/jenkins/1842)

### CODE REVIEW

**Comment by KostyaSha** [comment]
greedy https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/ removed artifacts

**Comment by KostyaSha** [comment]
https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/3228/
Test Result (no failures)

**Comment by jtnord** [comment]
I find this counter intuitive.
Give `save()` has been called I would expect the trigger to be correctly loaded and configured (after all the job is now on disk and configuration is fully visible in the UI)  - yet this is not done until `load()` is called - which seems to me to be incorrect.

**Comment by KostyaSha** [comment]
> I find this counter intuitive.
> Give save() has been called I would expect the trigger to be correctly loaded and configured (after all the job is now on disk and configuration is fully visible in the UI) - yet this is not done until load() is called - which seems to me to be incorrect.

This is not my logic, i also was confused by doing roundtrip for making job buildable like UI action do. This is miss behaviour in Job/Project <-> Trigger design. Is not something that i can change as i even can't read @jglick coding style :(

**Comment by jtnord** [comment]
```java
if (isBuildable()) {
    trigger.start(this, true);
}
```
in [this method](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/AbstractProject.java#L1578)?

may be a valid reason this is not done - but at least that method calls `save()`  :-o

**Comment by KostyaSha** [comment]
@jtnord note there is a terminology `isBuildable()` job can be saved but wouldn't be `buildable` i.e. after copy. Submit ui also calls trigger initialisation, so probably this logic should be in job or somewhere else (or already exists) and job should have `activate()` or `initialise`... I tried follow this logic but it also complicated because dependency calculation is related here.

**Comment by KostyaSha** [comment]
@jtnord this additional top logic is here 
https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/AbstractProject.java#L1815-L1881

**Comment by jtnord** [comment]
:+1: as the code fixes the issue and the fix is trivail - however I'm not convinced that this is not just masking something

**Comment by KostyaSha** [comment]
@jtnord the root issue that `job` in Trigger according to design can be null,  see https://issues.jenkins-ci.org/browse/JENKINS-29794  and this optimisation https://issues.jenkins-ci.org/browse/JENKINS-29792 to not do many redundant calss for polling and etc exposed it. 
When i implemented Trigger i got all NPEs during development and prepared trigger code, seems other implementers ignored random NPEs (btw there was many PRs about this null job before, but nobody fixed).
I also checked coverity scan for trigger classes and didn't found NPEs. Still waiting @oleg-nenashev to re-run new scan.

**Comment by jtnord** [comment]
@KostyaSha you had my :+1:

File: test/src/test/java/jenkins/triggers/ReverseBuildTriggerTest.java [hunk 1]
```diff
@@ -148,4 +157,30 @@
...
 
+    @Issue("JENKINS-29876")
+    @Test
+    public void nullJobInTriggerNotCausesNPE() throws Exception {
+        final FreeStyleProject upstreamJob = r.createFreeStyleProject("upstream");
+
+        //job with trigger.job == null
+        final FreeStyleProject downstreamJob1 = r.createFreeStyleProject("downstream1");
+        final ReverseBuildTrigger reverseBuildTrigger = new ReverseBuildTrigger("upstream", Result.SUCCESS);
+        downstreamJob1.addTrigger(reverseBuildTrigger);
```
Comment (jglick) [inline test/src/test/java/jenkins/triggers/ReverseBuildTriggerTest.java]: Hmm, does `addTrigger` not set the `job`?!

File: core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java [hunk 2]
```diff
@@ -137,7 +137,8 @@ private boolean shouldTrigger(Run upstreamBuild, TaskListener listener) {
         return result != null && result.isBetterOrEqualTo(threshold);
     }
 
-    @Override public void buildDependencyGraph(final AbstractProject downstream, DependencyGraph graph) {
+    @Override
+    public void buildDependencyGraph(final AbstractProject downstream, DependencyGraph graph) {
```
Comment (jglick) [inline core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java]: Gratuitous whitespace-only changes should be avoided to minimize work with cherry-picking etc.

**Comment by jglick** [comment]
:+1:

File: test/src/test/java/jenkins/triggers/ReverseBuildTriggerTest.java (hunk 1 above)
Comment (KostyaSha) [inline test/src/test/java/jenkins/triggers/ReverseBuildTriggerTest.java]: No, this confused me during tests implementation. I thought that's why you did roundtrips in other test cases :)
see https://github.com/jenkinsci/jenkins/pull/1842#issuecomment-144721166

File: core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java (hunk 2 above)
Comment (KostyaSha) [inline core/src/main/java/jenkins/triggers/ReverseBuildTrigger.java]: @jglick say the same here https://github.com/jenkinsci/jenkins/pull/1815

**Comment by KostyaSha** [comment]
Restored @jglick coding style and seems it added one annotation that i played with. Looks right, so i will keep it.

**Comment by oleg-nenashev** [comment]
:+1:

File: test/src/test/java/jenkins/triggers/ReverseBuildTriggerTest.java (hunk 1 above)
Comment (KostyaSha) [inline test/src/test/java/jenkins/triggers/ReverseBuildTriggerTest.java]: @jglick filled https://issues.jenkins-ci.org/browse/JENKINS-30802 as it really confuses everybody

**Pull request merged**